File: FSharpConsoleApp/HappyPathCodingBasics.fs
module FSharpConsoleApp.HappyPathCodingBasics

open FsToolkit.ErrorHandling

let asynchronous id =
    async {
        return id + 1
    }

let caller id =
    async {
        let! r = asynchronous id
        return r
    }

let optional id = if id % 2 = 0 then Some id else None

let optionalPlus17 id =
    option {
        let! r = optional id
        return r + 17
    }

let getResult id =
    if id % 2 = 0 then Ok id
    else Error "odd number"

let resultPlus17 id =
    result {
        let! r = getResult id
        return r + 17
    }

let callResult () =
    let r = getResult 42
    match r with
    | Ok value -> "value"
    | Error error -> "error"

File: CSharpConsoleApp/HappyPathCodingBasics.cs
using System.Threading.Tasks;

namespace CSharpConsoleApp
{
    public class HappyPathCodingBasics
    {
        public async Task<int> Asynchronous(int id)
        {
            return id + 1;
        }

        public async Task<int> Caller(int id)
        {
            var r = await Asynchronous(id);
            return r;
        }

        public int? Optional(int id)
        {
            var r =
                id % 2 == 0 ?
                    (int?)id :
                    null;
            return r;
        }

        public (int?, string?) Result(int id)
        {
            var r = id % 2 == 0 ?
                ((int?)id, null) :
                (default(int?), "odd number");
            return r;
        }

        public async Task<(int?, string?)> AsyncResult(
            int id)
        {
            var async =
                await Asynchronous(id).ConfigureAwait(false);

            var optional = Optional(id);
            if (optional == null)
            {
                return (null, "no value");
            }

            var result = Result(id);
            if (result.Item1 == null)
            {
                return result;
            }

            return
                (
                    async + optional.Value + result.Item1.Value,
                    null
                );
        }
    }
}